awscli/formatter.py

```diff
@@ -120,17 +128,70 @@ def _format_response(self, command_name, response, stream):
             # - the json dumper will complain if you pass them in. datetime
             # values should respect the cli timestamp format, which is
             # impossible to do from the Formatter.
-            json.dump(response, stream, ensure_ascii=False)
+            json.dump(value, stream, ensure_ascii=False)
             stream.write('\n')
         else:
-            self._yaml.dump(response, stream)
+            self._yaml.dump(value, stream)
 
     def _is_json_scalar(self, value):
         if value is None:
             return True
         return isinstance(value, (int, float, bool, compat.six.string_types))
 
 
+class YAMLFormatter(FullyBufferedFormatter):
+    def __init__(self, args, yaml_dumper=None):
+        super(YAMLFormatter, self).__init__(args)
+        self._yaml_dumper = yaml_dumper
+        if yaml_dumper is None:
+            self._yaml_dumper = YAMLDumper()
+
+    def _format_response(self, command_name, response, stream):
+        if response == {}:
+            return None
+        self._yaml_dumper.dump(response, stream)
+
+
+class StreamedYAMLFormatter(Formatter):
+    def __init__(self, args, yaml_dumper=None):
+        super(StreamedYAMLFormatter, self).__init__(args)
+        self._yaml_dumper = yaml_dumper
+        if yaml_dumper is None:
+            self._yaml_dumper = YAMLDumper()
+
+    def __call__(self, command_name, response, stream=None):
+        if stream is None:
+            stream = self._get_default_stream()
+        response_stream = self._get_response_stream(response)
+        for response in response_stream:
+            try:
+                # For YAML it is ambiguous as to whether the output from the
+                # stream is N responses in 1 list or N lists each with 1
+                # response. We go with the latter so we can reuse our YAML
+                # dumper
+                self._yaml_dumper.dump([response], stream)
+            except IOError:
+                # If the reading end of our stdout stream has closed the file
+                # we can just exit.
+                return
+            finally:
+                # flush is needed to avoid the "close failed in file object
+                # destructor" in python2.x. See:
+                # http://bugs.python.org/issue11380).
+                self._flush_stream(stream)
+
+    def _get_response_stream(self, response):
+        if is_response_paginated(response):
+            return map(self._get_transformed_response_for_output, response)
```

Comment:
Interesting.  While technically this doesn't affect our v2 builds because we use python3, this does technically change the semantics of a streaming response in python2 because `map()` isn't lazy in py2.  It's an interesting question, but I'm inclined to say that as long as we still test on python2.7, we should try to preserve behavior where possible.  What do you think about adding a something in `compat.py` and falling back to `itertools.imap` for py2?

Reply:
Good catch. I was not even thinking about Python 2 here. Yeah I'll add a compat for it.